Navigator: add support for exit animation

ciampo opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Part of #59418
Requires #64786 to be merged first

Add exit animations to `Navigator` screen

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

An exit animation can help the user understand the transition to a new screen better — and gives an additional sense of polish to the UI.

The animation specs (durations, delays, easing functions) are also updated following the latest motion guidelines (cc @nick-a8c)

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

  - Added a `useScreenAnimatePresence` hook which takes care of managing the screen's animation lifecycle
  - Use CSS transitions to animate the screen
  - Use `data-` attributes to set animation state on the DOM element, which is used to trigger the CSS transitions
  - Use CSS grid on the provider to force all screens to overlap, while also pushing the height of the container
  - Always render the screen root element, while conditionally rendering its contents
  - Updated tests to completely skip animations, that were causing errors and general flakiness
  - Added a note in the README about setting the wrapper height (related conversation in https://github.com/WordPress/gutenberg/pull/64798#discussion_r1733235033)


## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

- Make sure that the animations are always applied to the pair of screens (the one transitioning in and the one transitioning out), both when navigating to a new screen and back to the parent screen;
- Make sure that the animations work as expected on all supported browsers;
- Make sure that the animations are not performed when enabling "reduced motion" browser/OS settings


### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

https://github.com/user-attachments/assets/bee8f934-0f0c-4df1-93ae-5510b2f4abd9

10x slow-motion:

https://github.com/user-attachments/assets/d802adc5-6b4f-4803-b8c0-26a421e618cf

ciampo committed:
Refactor screen logic for better clarity and hook deps

ciampo committed:
Add exit animation, rewrite screen animations / DOM rendering logic

ciampo committed:
Only add inset CSS rule when animating out

ciampo committed:
Parametrise animation

ciampo committed:
CHANGELOG

jasmussen commented:
Nice. Is it possible to test this in the site editor? It's not clear that the new animation from this branch is applied there:

![nav](https://github.com/user-attachments/assets/79390517-e3ce-4bc5-a58f-8295ffab27b9)

Here's trunk for comparison:

![trunk](https://github.com/user-attachments/assets/eadc8251-b524-4000-a889-7a09c960200f)

ciampo commented:
It appears that the site editor sidebar doesn't use `Navigator` anymore — looking at the file history, the change happened in https://github.com/WordPress/gutenberg/pull/60466. I was on parental leave at the time, but it seems like the main reason was to use the global site-editor router instead of having to coordinate with `Navigator`'s internal router.

`Navigator` is still being used in the Global Styles sidebar, though.

ciampo commented on packages/components/src/navigator/navigator-screen/component.tsx:
```diff
@@ -122,20 +181,40 @@ function UnconnectedNavigatorScreen(
+		<View
+			ref={ mergedWrapperRef }
+			className={ classes }
+			onAnimationEnd={ ( e: AnimationEvent ) => {
+				if ( ! isMatch && isExitAnimation( e ) ) {
+					// When the exit animation ends on an unselected screen, set the
+					// status to 'animated' to remove the screen contents from the DOM.
+					setExitAnimationStatus( 'animated' );
+				}
+			} }
```
I may consider adding a timeout to make sure that, if for whatever reason the correct `animationend` event doensn't fire, the timeout is going to still cause the screen to be correctly from the DOM.

ciampo replied:
Added in https://github.com/WordPress/gutenberg/pull/64777/commits/0ae892a27f7ea9dee23efd8fd968847dec0fe078

ciampo committed:
Add fallback timeout

mirka reviewed (commented):
Animating presence will likely be a recurring need. How can we abstract this into something encapsulated and reusable, so we don't have to add in a bunch of ad hoc animation coordinating logic into a component every time we need an entry/exit animation?

Not immediately sure if we should use a library like `react-transition-group` ([maybe](https://bundlephobia.com/package/react-transition-group@4.4.5) 1.7kB gzipped if we only needed the `Transition` component), or if we want to maintain our own for an even smaller size. Thoughts?

mirka commented on packages/components/src/navigator/navigator-screen/component.tsx:
```diff
@@ -29,21 +31,47 @@ import { NavigatorContext } from '../context';
 import type { NavigatorScreenProps } from '../types';
 
+const isReducedMotion = ( w: Window | null | undefined ) =>
+	!! w && w.matchMedia( `(prefers-reduced-motion)` ).matches === true;
```
I found a [`useReducedMotion` hook](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-compose/#usereducedmotion) in wp-compose.

ciampo replied:
Good suggestion! Done in https://github.com/WordPress/gutenberg/pull/64777/commits/15fe373ca8e090494f5190874ec8daad91f12e6f

ciampo committed:
Mention wrapper height in README

mirka reviewed (commented):
> I wonder if doing so would represent a form of early optimization. Maybe we can keep `Navigator` as-is, and re-discuss this topic if/when we have more real use cases for animating presence, so that we can better understand what type solution we'd need to implement?

I'm not familiar enough to make a confident judgement in either direction, but I trust you on your assessment after working on it.

I guess what immediately makes me a bit uncomfortable though is how the animation logic is interspersed in the component code. Do you think it would be worth the effort to extract most or all of the animation logic into a single hook? No regard for reusability at the moment, but just to isolate all the animation concerns into a single place. I'm already a bit afraid of touching this code, and I think separating the concerns would help a lot to ease that sense of complexity.

ciampo commented:
Rebased to include all recent `Navigator` changes. This PR is ready for another round of review

ciampo committed:
Use `useReducedMotion()` hook instead of custom logic

ciampo committed:
Extract useScreenAnimatePresence hook, tidy up

tyxla commented on packages/components/src/navigator/navigator-provider/README.md:
```diff
@@ -47,6 +47,10 @@ For example:
 -   `/parent/:param` is a child of `/parent` as well.
 -   if the current screen has a `path` with value `/parent/child/grand-child`, when going "back" `Navigator` will try to recursively navigate the path hierarchy until a matching screen (or the root `/`) is found.
 
+### Height and animations
+
+Due to how `NavigatorScreen` animations work, it is recommended that the `NavigatorProvider` component is given enough height to match the tallest `NavigatorScreen`. Not doing so could result in glitchy animations, especially when transitioning from a taller to a shorter `NavigatorScreen`.
```
Ah, this is a bummer. Looks like something the component should be able to do on its own. Is there a reason why we can't automate this, using a resize observer for example?

ciampo replied:
I pushed some more work that should greatly improve this aspect.

ciampo commented:
> I guess what immediately makes me a bit uncomfortable though is how the animation logic is interspersed in the component code. Do you think it would be worth the effort to extract most or all of the animation logic into a single hook? No regard for reusability at the moment, but just to isolate all the animation concerns into a single place.

This is a great suggestion, I definitely agree. I pushed a couple of commits and things are looking much more clear :)

ciampo commented on packages/components/src/navigator/navigator-screen/component.tsx:
```diff
@@ -53,30 +65,26 @@ function UnconnectedNavigatorScreen(
+		[ className, cx, animationStyles ]
 	);
 
+	// Focus restoration
```
There are no runtime changes re. focus restoration, only variable renaming and some prop destructuring that happened outside the `useEffect` hook

tyxla commented on packages/components/src/navigator/test/index.tsx:
```diff
@@ -477,16 +478,27 @@ describe( 'Navigator', () => {
+	// Mock `matchMedia` so that all animations are skipped.
+	const mockedMatchMedia = jest.fn( ( query: string ) => {
+		if ( /prefers-reduced-motion/.test( query ) ) {
+			return { matches: true } as ReturnType< typeof window.matchMedia >;
```
Isn't it a better bet to test with all animations enabled? After all, there might be bugs that are only encountered with animations on.

ciampo replied:
I don't think `js-dom` supports animations well — `testing-library` has even an [example](https://testing-library.com/docs/example-react-transition-group/#mock) of how to mock `React Transition Group`.

tyxla reviewed (commented):
I still need to test and review this thoroughly, but I'm submitting some early, naive feedback.

tyxla commented on packages/components/src/navigator/navigator-screen/component.tsx:
```diff
@@ -28,22 +27,35 @@ import { View } from '../../view';
 		'NavigatorScreen'
 	);
 
+	// Read navigator context, destructure location props.
```
Some of the comments seem a bit redundant since the code is self-explicatory. Were you planning on cleaning them up once you're done?

ciampo replied:
Removed obvious comments

ciampo committed:
Forward animationEnd

tyxla commented on packages/components/src/navigator/navigator-screen/use-screen-animate-presence.ts:
```diff
@@ -0,0 +1,122 @@
+const isExitAnimation = ( animationName: string ) =>
+	animationName === styles.slideToLeft.name || styles.slideToRight.name;
+
+export function useScreenAnimatePresence( {
```
Lots of good work here, but I'm wondering if we're trying too hard not to use `framer-motion`. When it's about a simple animation, using CSS makes lots of sense. But when we need to reinvent the wheel in cases like animating presence, I feel like the best bet would be to use `framer-motion`. Bundle size is important, but DX and UX are more important, so we might have to make compromises. Especially given that we're looking to build more complex motion and animations for many of the components and flows.

mirka replied:
I definitely feel _better_ about it after the refactor into a designated hook, but also it’s still unknown how bug-free the current code is, and how many other components will need to animate presence.

My instinct is that it could still make sense to migrate from Framer Motion to a lighter library like React Transition Group. But then again, if it's a substantial effort to migrate from FM to RTG, maybe even that isn't worth it. 

I’d love to hear from @ciampo and @daniguardiola about how they currently assess the overall cost/benefit around these rewrites. Let's make sure the benefits are still big enough to justify our rewrite efforts and potential increase in maintenance cost.

mirka replied:
As for this particular PR, I'm personally comfortable enough with the logic isolation now to merge it as an incremental improvement over trunk. Is it scalable/sustainable? Maybe not. But it seems like we could swap it out without too much difficulty if we ultimately decide to use a library.

tyxla replied:
I don't have particularly strong feelings about blocking this particular PR in favor of `framer-motion`, but I share the sentiment on scalability, the extra maintenance cost, and moving in that direction in the future based on the amount of custom work and wheel rediscovery we need to do. It's worth considering doing more specific measurements to ensure that a refactor away from `framer-motion` actually yields noticeable results. We should also consider the opportunity cost - what we could have built/shipped if we reused `framer-motion` instead of rediscovering it and maintaining the extra layer of complexity.

ciampo committed:
Add `setWrapperHeight` functionality via context

ciampo committed:
Use `clip` instead of `hidden` for overflow-x

ciampo committed:
Less aggressive clipping for screens that are animating out

ciampo committed:
Better sizing styles for screen, to keep it more stable while transitioning out

ciampo committed:
Refine internal animation logic for less jumpy animations

ciampo committed:
Remove unnecessary Storybook styles

ciampo committed:
Change wording

ciampo commented:
I've pushed some further updates to the logic, making it (hopefully) clearer and making more resilient to UI "jumping" while switching across screens.

I guess the conversation started in https://github.com/WordPress/gutenberg/pull/64777#discussion_r1736202491 will inform the future of this PR — whether we go ahead with the current approach, or if we revert back to using a higher-level animation library.

ciampo commented:
Pausing the effort on this PR for now for a couple of reasons:

- I'd like to have agreement on [whether we should use a 3rd party library or not](https://github.com/WordPress/gutenberg/pull/64777#discussion_r1736202491)
- some animations in the global styles sidebar are _slightly_ glitchy for reasons I yet don't understand (especially when going from the Typography screen to the individual Element (ie. Text, Headings, Lists...) typography screen. I won't spend further time debugging the issue if we're not going to go down this PR's route

ciampo committed:
Improve logic:

- more clear animation status names
- apply CSS animation only while effectively animating
- fix bug in the animationEnd callback which was matching animation end
  events too loosely, thus causing glitches

ciampo commented:
I had a couple of ideas going around in my mind, I tested them and they seem to improve the logic.

I also discovered this small "transition state machine" library that could delegate some logic to a third party, if we even think it's worth exploring. https://www.npmjs.com/package/react-transition-state

jsnajdr commented on packages/components/src/navigator/navigator-screen/use-screen-animate-presence.ts:
```diff
@@ -0,0 +1,173 @@
+	// `animationend` event is not triggered.
+	useEffect( () => {
+		if ( animationStatus === 'ANIMATING_OUT' ) {
+			animationTimeoutRef.current = window.setTimeout( () => {
```
`animationTimeout` can be a local variable in the effect function, no need for an external ref.

Also, should this fallback timeout also change the state from `ANIMATING_IN` to `IN`? The `onScreenAnimationEnd` does that. If I understand correctly, the idea is that if the `animationEnd` callback doesn't run for some reason, the fallback callback will do its job anyway after some timeout.

ciampo replied:
Yes, that's correct — I forgot to update the timeout in the last refactor

ciampo replied:
https://github.com/WordPress/gutenberg/pull/64777/commits/ce9e38acbc9ddb274bcfb6ecf61277fa1e8987bf

ciampo committed:
Add timeout fallback for in animations too

ciampo committed:
Fix animation delay for forwards.out animation

jsnajdr commented:
Here's how I understand the concept:

The `NavigatorProvider` no longer renders just one matching `NavigatorScreen`, but it can render two or even more. One is the real active one, the others are there only for effect. We could build a navigator that displays all parent screens using a `wasMatch` array, like this Apple Time Machine history UI 🙂

![ea02d34d-0309-4109-819a-e05eb5087624](https://github.com/user-attachments/assets/f0e46860-93c8-4589-9dfc-fa33cc76b8ce)

So, that determines the DOM structure and the baseline styles for navigator, even without any animations. The matching screens are displayed side-by-side, or they are all `display:none` except one, or are displayed as something else. The styles must be able to show the active screen as the active one and the other screens as inactive.

I don't like the `setWrapperHeight` helper very much. It feels very hacky. Instead, the base navigator styles should be able to display multiple screens at once without any helpers. When I remove `setWrapperHeight` from the code, I don't see any ugly effects in the Storybook examples.

There can be a behavior that the `wasMatch` value is reset to `false` after some timeout and the inactive screen is unmounted. That will be useful when the exit animation is implemented. But it's an optimization, not an essential feature. The navigator still works 100% when the `wasMatch` screen is rendered indefinitely.

When the base DOM structure and the base styles are established, the animations are merely an icing on the cake. The transition from active screen to inactive and vice versa are animated. Can be easily done with CSS animations.

I think that when you think about the PR as i) solid base DOM structure and styles + ii) simple animations on top, it's easier to understand.

And if we did this with `framer-motion`, it doesn't give us much extra, does it? Its `AnimatePresence` helper is useful, but that's all.

ciampo committed:
Use display: grid instead of absolute positioning to set provider min height

Remove unnecessary import

ciampo commented:
> When I remove setWrapperHeight from the code, I don't see any ugly effects in the Storybook examples.

That's because, in the storybook examples, the `NavigatorProvider` component has a `height: calc(100vh - 2rem)` style applied.

If you remove the set `height` and remove `setWrapperHeight` too, you should notice that the screen animating out will have its contents cutoff if the screen animating in is not as tall.

Here is a video comparison (I slowed down the animation and added a background gradient to make things more evident)

| Without `setWrapperHeight` | With `setWrapperHeight` |
|---|---|
| <video src="https://github.com/user-attachments/assets/6d4d30af-1d0b-4a88-a8c9-300823ca2752" /> | <video src="https://github.com/user-attachments/assets/6658001e-2698-40cd-a2d8-02e7f0481825" /> |

Having said that, this comment made me reflect:

> I don't like the setWrapperHeight helper very much. It feels very hacky. Instead, the base navigator styles should be able to display multiple screens at once without any helpers.

And I realised that, instead of setting `position: absolute` on the screen transitioning out (which then causes the `NavigatorProvider` to jump, and the screen animating out to be possible cut off), we can use `display: grid` and have both screens (the one transitioning in, and the one transitioning out) overlap in the same grid cell, meaning that the tallest one will push the cell's height without the need for `setWrapperHeight`.

I applied this approach in [`4bbb6db` (#64777)](https://github.com/WordPress/gutenberg/pull/64777/commits/4bbb6db875058dd9d8d7582506731b3ac82df1d1)

ciampo commented:
> Here's how I understand the concept:
> 
> The `NavigatorProvider` no longer renders just one matching `NavigatorScreen`, but it can render two or even more. One is the real active one, the others are there only for effect. We could build a navigator that displays all parent screens using a `wasMatch` array, like this Apple Time Machine history UI 🙂
> 
> So, that determines the DOM structure and the baseline styles for navigator, even without any animations. The matching screens are displayed side-by-side, or they are all `display:none` except one, or are displayed as something else. The styles must be able to show the active screen as the active one and the other screens as inactive.


To be honest, this is very similar to how `Navigator` works.

The `NavigatorProvider` keeps all the logic, and lets every screen know what's the selected screen ID. So, every `NavigatorScreen` is always rendered (in the react sense).

Each screen, then, decides independently if its contents should be rendered in the DOM, and handles focus if/when needed.

> When the base DOM structure and the base styles are established, the animations are merely an icing on the cake. The transition from active screen to inactive and vice versa are animated. Can be easily done with CSS animations.
> I think that when you think about the PR as i) solid base DOM structure and styles + ii) simple animations on top, it's easier to understand.

Yup, that's basically what's happening with the `useScreenAnimatePresence` logic.

The hook is basically a state machine that goes through the "animation lifecycle" of a navigator screen:

```
initial     →           animating in          →        visible
                             ↓                            ↓
                   removed from the DOM       ←      animating out
```     

> And if we did this with `framer-motion`, it doesn't give us much extra, does it? Its `AnimatePresence` helper is useful, but that's all.

Framer motion would basically replace the `useScreenAnimatePresence` hook, but everything else would still need to be kept around.

ciampo committed:
Simplify navigatorScreenAnimation

ciampo commented:
On a different note, the bug that I mentioned before in the global styles sidebar continues to persist. I tried to slow down the animations in the dev tools, but that causes the bug not to be reproducible. I had to use my phone to record a slow-motion video of the transition:


https://github.com/user-attachments/assets/9b9972b0-c570-48e3-81cf-1703ed771eb8

It's as if the paragraph screen was selected, then deselected for a brief moment, and then selected again.

I will look into it, as I'm not sure yet if this is something related to how `Navigator` is used in global styles, or if this is an internal bug to `Navigator`.

ciampo committed:
Remove unnecessary state element ref

ciampo committed:
Use "start" and "end" instead of "backwards" and "forwards"

ciampo committed:
Do not rely on `usePrevious`

ciampo committed:
Use CSS transitions

ciampo committed:
Fix Storybook example

ciampo commented:
So!

I spent a lot more time than I would have liked to debug [the glitch in the global styles sidebar](https://github.com/WordPress/gutenberg/pull/64777#issuecomment-2334381574), and couldn't really get to the bottom of it. Somehow, the CSS animation gets re-triggered,and I wasn't able to understand the reason (not sure if @jsnajdr wants to attempt to debug it too).

I therefore decided to apply a different animation strategy, **using CSS transitions instead of animations** since they don't restart if the underlying styles get re-applied. I tested the new approach in Storybook and in the editor, both Global Styles sidebar and preferences modal (on mobile viewport) and it seems to work well.

@WordPress/gutenberg-components , do we think we can polish + merge the current approach? (I also made a few changes to the global styles sidebar code, which I will extract to separate PRs if we go forward with the current set of changes)

ciampo commented on packages/components/src/navigator/navigator-screen/use-screen-animate-presence.ts:
```diff
@@ -0,0 +1,158 @@
+
+	const { animationStatus, animationDirection } = animationState;
+
+	// Start enter and exit animations when the screen is selected or deselected.
+	// The animation status is set to `*_END` immediately if the animation should
+	// be skipped.
```
This comment is out of date, I'll remove it in a future commit

ciampo committed:
Revert "Use CSS transitions"

This reverts commit 946f9c953232b788f58050d2a94d9d131527a180.

ciampo committed:
Switch to data-attributes for less runtime emotion calculations

ciampo committed:
Add back fallback animation timeout

ciampo committed:
Move CHANGELOG entry to unreleased section

ciampo commented:
With https://github.com/WordPress/gutenberg/pull/65494 merged, the glitch reported above should be fixed (thank you @jsnajdr 🙏 )

I rebased, restored the CSS animations based approach, and added a few tweaks (like moving to `data-*` attributes, instead of dynamic emotion styles).

I'm going to extract the changes to the `@wordpress/edit-site` package to separate PRs, but in the meantime this PR is ready to be tested again.

ciampo committed:
Clean up code, reduce diff for easier reviewing

ciampo commented:
With https://github.com/WordPress/gutenberg/pull/65494 merged, the [glitch](https://github.com/WordPress/gutenberg/pull/64777#issuecomment-2334381574) reported above should be fixed (thank you @jsnajdr 🙏 )

I rebased, restored the CSS animation- based approach, and added a few tweaks (like moving to `data-*` attributes, instead of dynamic emotion styles).

I'm going to extract the changes to the `@wordpress/edit-site` package to separate PRs (#65522, #65523, #65527), but in the meantime this PR is ready to be tested again.

ciampo commented:
> With #65494 merged, the [glitch](https://github.com/WordPress/gutenberg/pull/64777#issuecomment-2334381574) reported above should be fixed

Since that PR [was reverted](https://github.com/WordPress/gutenberg/pull/65533) I'd imagine the glitch would be back when rebasing on top of latest trunk 😭

jsnajdr commented:
> Since that PR was reverted I'd imagine the glitch would be back when rebasing on top of latest trunk

I'm trying again in #65564, this time with a much better and simpler implementation 🙂

tyxla commented:
Sounds like we should mark this one as blocked until #65564 lands.

ciampo commented:
With https://github.com/WordPress/gutenberg/pull/65564, https://github.com/WordPress/gutenberg/pull/65522, https://github.com/WordPress/gutenberg/pull/65523, and https://github.com/WordPress/gutenberg/pull/65527 merged, I rebased this PR and solved conflicts.

It tested well on my machine all all browsers

jsnajdr reviewed (approved):
I don't have any further feedback, everything has been addressed. The glitch with Typography panel is solved. I think this is ready to 🚢

t-hamano commented:
This PR seems to have caused problems with the shadow preset panel and font size preset panel.

Looking at the browser error log, it might be a problem with the implementation of the shadow/font size preset, but could you take a look at the following issue?

- https://github.com/WordPress/gutenberg/issues/65711

ciampo commented:
> This PR seems to have caused problems with the shadow preset panel and font size preset panel.
> 
> Looking at the browser error log, it might be a problem with the implementation of the shadow/font size preset, but could you take a look at the following issue?
> 
> * [Shadow/Font size preset panel crashes the editor #65711](https://github.com/WordPress/gutenberg/issues/65711)

Thank you for flagging — left a reply in the issue itself.

jeffpaul commented:
@nick-a8c would you mind sharing your WordPress.org username so we can ensure you're properly credited in the 6.7 release next week?
